### PR TITLE
Storm-Pulsar bolt and spout

### PR DESCRIPTION
**Pull request by dlg99**
## What is the purpose of the change

Apache Pulsar is a popular streaming and messaging platform. 
It provides storm integration as part of the [pulsar-adapters](https://github.com/apache/pulsar-adapters) subproject but the storm adapter never keeps up with the latest storm changes/builds. 
The adapter does not require much knowledge about the pulsar internals but knowledge of Storm internals is something that is more beneficial. 
This PR is an attempt to transition pulsar-storm adapter to the storm community .

## How was the change tested

Unit/integration test.

### CODE REVIEW

**Comment by dlg99** [comment]
Travis build is failing for unrelated reasons. I built and tested as:
```
mvn clean install -DskipTests -P '!examples,!externals'
cd external/storm-pulsar/
mvn clean install
...
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.pulsar.storm.PulsarSpoutIntegrationTest
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 71.711 s - in org.apache.pulsar.storm.PulsarSpoutIntegrationTest
[INFO] Running org.apache.pulsar.storm.PulsarSpoutTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.098 s - in org.apache.pulsar.storm.PulsarSpoutTest
[INFO] Running org.apache.pulsar.storm.PulsarBoltIntegrationTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 42.811 s - in org.apache.pulsar.storm.PulsarBoltIntegrationTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 22, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

**Comment by rzo1** [comment]
Travis needs https://github.com/apache/storm/pull/3437  ...

File: external/storm-pulsar/pom.xml [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+-->
+<project
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
+  xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+      <artifactId>storm</artifactId>
+      <groupId>org.apache.storm</groupId>
+      <version>2.4.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (dlg99) [reply external/storm-pulsar/pom.xml]: done

File: external/storm-pulsar/pom.xml [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  <dependencies>
+
+    <dependency>
+      <groupId>org.apache.storm</groupId>
+      <artifactId>storm-client</artifactId>
+      <version>${project.version}</version>
+      <scope>provided</scope>
+      <exclusions>
+        <exclusion>
+          <groupId>ch.qos.logback</groupId>
```
[the remark this answers is not in the document]
Comment (dlg99) [reply external/storm-pulsar/pom.xml:62]: The code is moved from https://github.com/apache/pulsar-adapters 
Why do we need to include the logback, wouldn't it be provided by the storm runtime anyway

File: external/storm-pulsar/src/test/java/org/apache/pulsar/storm/PulsarBoltIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+import org.apache.pulsar.client.api.ProducerConsumerBase;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.TypedMessageBuilder;
+import org.apache.pulsar.common.policies.data.TopicStats;
+import org.apache.storm.task.OutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.topology.OutputFieldsDeclarer;
+import org.apache.storm.tuple.Tuple;
+
+import org.junit.After;
```
Comment (bipinprasad) [inline external/storm-pulsar/src/test/java/org/apache/pulsar/storm/PulsarBoltIntegrationTest.java:38]: Can you change junit4 to junit5. We are trying to move away from junit4 and would like new code to be junit5.

**Comment by vesense** [comment]
Overall looks good to me. Thanks @dlg99 for the great job. And do you have any updates for @agresch and @bipinprasad comments?

**Review by agresch — APPROVED** [review]

**Pull request merged**